### PR TITLE
implement RFC 1194

### LINKED ISSUE #28050: Set recovery methods
Tracking issue for https://github.com/rust-lang/rfcs/pull/1194

cc https://github.com/rust-lang/rust/pull/28043

### PR DESCRIPTION
**Pull request by apasel422**
closes #28050

### CODE REVIEW

File: src/libstd/collections/hash/set.rs [hunk 1]
```diff
@@ -568,6 +588,18 @@ impl<T, S> HashSet<T, S>
     {
         self.map.remove(value).is_some()
     }
+
+    /// Removes and returns the value in the set, if any, that is equal to the given one.
+    ///
+    /// The value may be any borrowed form of the set's value type, but
+    /// `Hash` and `Eq` on the borrowed form *must* match those for
+    /// the value type.
+    #[unstable(feature = "set_recovery", issue = "0")]
```
Comment (alexcrichton) [inline src/libstd/collections/hash/set.rs]: (just to make sure this comment isn't lost)

These issues can be wired up to the tracking issue for the RFC now

**Comment by alexcrichton** [comment]
Thanks @apasel422! Was this also going to take care of `BTreeSet`? Or is that holding off for future changes?

**Comment by apasel422** [comment]
I just haven't gotten around to making the changes for BTreeSet. I'll
update this PR once I have.

**Comment by alexcrichton** [comment]
Ok, well in the meantime these look good to me :)

**Comment by apasel422** [comment]
Updated with correct issue numbers and `BTreeSet` methods.

File: src/libcollections/btree/map.rs [hunk 2]
```diff
@@ -1609,3 +1609,86 @@ impl<K: Ord, V> BTreeMap<K, V> {
...
+    fn get(&self, key: &Q) -> Option<&K> {
+        let mut cur_node = &self.root;
+        loop {
+            match Node::search(cur_node, key) {
+                Found(handle) => return Some(handle.into_kv().0),
+                GoDown(handle) => match handle.force() {
+                    Leaf(_) => return None,
+                    Internal(internal_handle) => {
+                        cur_node = internal_handle.into_edge();
+                        continue;
```
Comment (alexcrichton) [inline src/libcollections/btree/map.rs]: I think this `continue` can be omitted

Comment (apasel422) [inline src/libcollections/btree/map.rs]: This is actually copied from `BTreeMap::get`, so I opted to keep the code as similar as possible.

**Comment by alexcrichton** [comment]
@bors: r+ f9b63d39730740e002cc31a38a7a8a3a18d3f46d

Thanks @apasel422!

**Pull request merged**